tests/unit/test_config_provider.py

```diff
@@ -942,3 +943,172 @@ def test_resolve_auto_mode_imds_region_provider_connect_timeout(self):
             )
             mode = smart_defaults_factory.resolve_auto_mode('us-west-2')
             assert mode == 'standard'
+
+
+def _known_service_names_and_ids():
+    my_session = session.get_session()
+    loader = my_session.get_component('data_loader')
+    available_services = loader.list_available_services('service-2')
+
+    result = set()
+    for service_name in available_services:
+        model = my_session.get_service_model(service_name)
+        result.add((model.service_name, model.service_id))
+    return sorted(result)
+
+
+class TestCustomEndpointProvider(unittest.TestCase):
+    def assert_does_provide(
+        self,
+        service,
+        environ_map,
+        full_config_map,
+        scoped_config_map,
+        expected_value,
+    ):
+        chain = ConfiguredEndpointProvider(
+            scoped_config=scoped_config_map,
+            full_config=full_config_map,
+            client_name=service,
+            environ=environ_map,
+        )
+        value = chain.provide()
+        self.assertEqual(value, expected_value)
+
+    def test_does_provide_env_var(self):
+        self.assert_does_provide(
+            service="batch",
+            environ_map={'AWS_ENDPOINT_URL': 'from-env'},
+            full_config_map={},
+            scoped_config_map={},
+            expected_value='from-env',
+        )
```

Comment:
It looks like you're running this function a lot. We should have one test function that just uses `pytest.mark.parametrize` instead. It should probably also exist outside of a test class and just be a standalone function.

Reply:
These are how the other config provider test cases are structured; as I interpreted it to provide readable names for the specific cases that would otherwise need to be reverse engineered from the contents (since the testing is often to confirm that something is done in the right order of precedence).

The functional tests (in a separate PR) parameterize over all possible combinations of values provided for setting the endpoint URL.

Reply:
I wouldn't use precedence in our test suite as a reason to structure tests this way. In pretty much every case when we're writing an entire new group of tests, we use newer functional patterns with pytest and use parametrize to remove redundancy. There isn't a need to use individual test functions as a means to document behavior of the new provider, so long as the feature is well-documented.